Mage_Core_Model_Abstract: Fix rollback when Throwable is thrown in save/delete method

Sekiphp opened the pull request:
### Description (*)
When you throw some child of `Error` class (or `Error` class) in `_beforeSave` or in `_afterSave` methods - Magento do not rollback started transaction and it is provide error: 
![image](https://user-images.githubusercontent.com/9967016/109885311-9ad05480-7c7e-11eb-9b1f-9e3953f95b98.png)

PHP implementation:
`Error implements Throwable`
`Exception implements Throwable`

### Manual testing scenarios (*)

Consider this code:
```
class Namespace_Module_Model_Foo extends Mage_Core_Model_Abstract {
    protected function _beforeSave()
    {
        parent::_beforeSave();

        throw new Error('error from some specifiec reason');
    }
}

Mage::getModel('namespace_module/foo')->save();
```


### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All automated tests passed successfully (all builds are green)

LukasCesal commented:
Funny when you are upgrading to LTS and get another bunch of errors :-) thank you Luboš

J-Kostecka commented:
This might also solve PZ-4381 - products disappearing randomly after import

kiatng commented:
See also PR #1442

kiatng commented:
@Sekiphp First, as you said, you have already fixed this in your project. Others facing similar issue can likewise fix them in their respective projects, no issue there. Second, for maintaining OpenMage code base, we need to fix `Throwable` for all instances, this helps to maintain code quality in terms of consistency, eg, use of `$t` or `$e`.  We can better progress OpenMage this way. I hope you agree.

colinmollenhour reviewed (changes requested):
Good change! Thanks for the PR!

nitpick: Please fix the vardoc and `delete()` as well while you're at it.

Sekiphp commented:
@colinmollenhour Added throws annotation and fixed `delete()` method

+ I changed `Throwable $t` to `Throwable $e`, because I read the specifications, and in official PHP examples is `$e` instead of `$t`:
- https://wiki.php.net/rfc/throwable-interface
- https://github.com/php/php-src/pull/1284
